src/Eloquent/Builder.php

```diff
@@ -202,56 +198,17 @@ public function raw($value = null)
         return $results;
     }
 
-    /**
-     * Attempt to create the record if it does not exist with the matching attributes.
-     * If the record exists, it will be returned.
-     *
-     * @param  array $attributes The attributes to check for duplicate records
-     * @param  array $values     The attributes to insert if no matching record is found
-     */
-    public function createOrFirst(array $attributes = [], array $values = []): Model
+    public function createOrFirst(array $attributes = [], array $values = []): Builder|Model
     {
-        if ($attributes === [] || $attributes === ['_id' => null]) {
-            throw new InvalidArgumentException('You must provide attributes to check for duplicates. Got ' . json_encode($attributes));
-        }
-
-        // Apply casting and default values to the attributes
-        // In case of duplicate key between the attributes and the values, the values have priority
-        $instance = $this->newModelInstance($values + $attributes);
-
-        /* @see \Illuminate\Database\Eloquent\Model::performInsert */
-        if ($instance->usesTimestamps()) {
-            $instance->updateTimestamps();
-        }
-
-        $values = $instance->getAttributes();
-        $attributes = array_intersect_key($attributes, $values);
-
-        return $this->raw(function (Collection $collection) use ($attributes, $values) {
-            $listener = new FindAndModifyCommandSubscriber();
-            $collection->getManager()->addSubscriber($listener);
-
-            try {
-                $document = $collection->findOneAndUpdate(
-                    $attributes,
-                    // Before MongoDB 5.0, $setOnInsert requires a non-empty document.
-                    // This should not be an issue as $values includes the query filter.
-                    ['$setOnInsert' => (object) $values],
-                    [
-                        'upsert' => true,
-                        'returnDocument' => FindOneAndUpdate::RETURN_DOCUMENT_AFTER,
-                        'typeMap' => ['root' => 'array', 'document' => 'array'],
-                    ],
-                );
-            } finally {
-                $collection->getManager()->removeSubscriber($listener);
+        try {
+            return $this->create(array_merge($attributes, $values));
+        } catch (WriteException $e) {
+            if ($e->getCode() === self::DUPLICATE_KEY_ERROR) {
```

Comment:
This approach seems untenable if any DuplicateKeyError will cause the transaction to be aborted.

Do I assume correctly that the library isn't creating its own transaction here, and it's only an issue when the application calls this _within_ its own transaction (as demonstrated in https://github.com/mongodb/laravel-mongodb/issues/2720)?

I think that in order to avoid DuplicateKeyError potentially aborting, we'd need to flip this to the original design of `firstOrCreate()` and only insert if the initial query found no match. That wouldn't eliminate the possibility of a write conflict between _two_ transactions, though.

Reply:
Alternatively, maybe you can check if you're already in a transaction and only then resort to an initial `find()` to try and avoid a DuplicateKeyError.

You could argue that two transactions conflicting with one another is a more general situation that users should anticipate.

Reply:
Some events (`saving` and `creating`) are triggered by `createOrFirst` but not by `firstOrCreate` when the document is found.

Reply:
I implemented your suggestion. When in transaction, `createOrFirst` fallback to `firstOrCreate`.